R/git-default-branch.R

```diff
@@ -0,0 +1,408 @@
+  ui_info("
+    Source repo is {ui_value(tr$repo_spec)} and its current default branch is \\
+    {ui_value(old_source_db)}.")
+
+  if (!isTRUE(tr$can_admin)) {
+    ui_stop("
+      You don't seem to have {ui_field('admin')} permissions for \\
+      {ui_value(tr$repo_spec)}, which is required to rename the default \\
+      branch.")
+  }
+
+  old_name <- old_name %||% guess_local_default_branch()
+
+  if (old_name != old_source_db) {
+    ui_oops("
+      It's weird that the current default branch for your local repo and \\
+      the source repo are different:
+      {ui_value(old_name)} (local) != {ui_value(old_source_db)} (source)")
+    if (ui_nope(
+      "Are you sure you want to proceed?",
+      yes = "yes", no = "no", shuffle = FALSE)) {
+      ui_stop("Cancelling.")
+    }
+  }
+
+  if (new_name == old_source_db) {
+    ui_info("
+      Source repo {ui_value(tr$repo_spec)} already has \\
+      {ui_value(old_source_db)} as its default branch.")
+  } else {
+    gh <- gh_tr(tr)
+    ui_done("
+      Renaming {ui_value(old_source_db)} branch to {ui_value(new_name)} in \\
+      the source repo {ui_value(tr$repo_spec)}.")
+    gh(
+      "POST /repos/{owner}/{repo}/branches/{from}/rename",
+      from = old_source_db,
+      new_name = new_name
+    )
+  }
+
+  rediscover_default_branch(old_name = old_name, verbose = FALSE)
+}
+
+git_default_branch_remote <- function(remote = "origin") {
+  url <- git_remotes()[[remote]]
+  if (length(url) == 0) {
+    ui_stop("There is no remote named {ui_value(remote)}.")
+  }
+
+  # TODO: generalize here for GHE hosts that don't include 'github'
+  parsed <- parse_github_remotes(url)
+  if (grepl("github", parsed$host)) {
+    return(github_remotes(remote, github_get = TRUE)$default_branch)
+  }
+
+  repo <- git_repo()
+  res <- tryCatch(
+    {
+      gert::git_fetch(remote = remote, repo = repo, verbose = FALSE)
+      gert::git_remote_ls(remote = remote, verbose = FALSE, repo = repo)
+    },
+    error = function(e) NA_character_
+  )
+  if (is.data.frame(res)) {
+    res <- path_file(res$symref[res$ref == "HEAD"])
+  }
+  res
+}
+
+guess_local_default_branch <- function(ask = is_interactive()) {
+  repo <- git_repo()
+
+  gb <- gert::git_branch_list(local = TRUE, repo = repo)[["name"]]
+  if (length(gb) == 0) {
+    ui_stop("
+      Can't find any local branches.
+      Do you need to make your first commit?")
+  }
+
+  if ("main" %in% gb && !"master" %in% gb) {
+    propose <- "main"
+  } else if ("master" %in% gb && !"main" %in% gb) {
+    propose <- "master"
+  } else if (length(gb) == 1) {
+    propose <- gb
+  } else {
+    ui_stop("
+      Not clear which existing local branch plays the role of the default.
+      You'll need to specify {ui_code('old_name')} explicitly.")
+  }
+
+  if (!ask || !is_interactive()) {
+    ui_done("
+      Local branch {ui_value(propose)} appears to play the role of \\
+      the default branch.")
+    return(propose)
+  }
+
+  if (ui_yeah("
```

Comment:
Hadley thinks I'm being too conservative and could proceed without this user confirmation. I see that point, but also kind of want the user to sign off before we plunge ahead.